Source/Orts.Simulation/Common/Scripting/TrainControlSystem.cs

```diff
@@ -560,4 +581,12 @@ public enum MonitoringStatus
         /// </summary>
         Intervention,
     }
+
+    public struct SignalFeatures
+    {
+        public string MainHeadSignalTypeName;
+        public Aspect Aspect;
+        public float DistanceM;
+        public float SpeedLimit;
```

Comment:
`SpeedLimit` needs SI units here.

Reply:
Also, script writers might be interested in information on the signal heads. Might want to consider including that into this struct.

Reply:
I suggested a similar struct with information about signal heads [http://www.elvastower.com/forums/index.php?/topic/33693-expand-potential-of-tcs-scripting/page__view__findpost__p__262321](here), but Peter thought this would be bad for performance. Using a fixed-size array or something for signal heads might force passing SignalFeatures by the stack, but I don't know if that will happen in C#.

Reply:
Will add the SI unit on SpeedLimit.
I am of the same opinion of Peter about not generating a heavier Struct (which by the way wouldn't be optimized for speedposts and so on), considering also that my small experience says that most of the calls don't need such info. Maybe a dedicated call like GetExtendedSignalInfo could be created, passing info about all heads, but that could be added later with no  problems.

Reply:
I disagree with Peter on the performance consideration; I would rather have an API that's easier to use. If building and destroying these structs takes a noticeable amount of time, then there are ways to avoid those operations by reusing the same struct across multiple calls.

That said, the way I would accomplish that would be use a class with public getters and internal setters instead of a struct, but we can handle these micro-optimizations later once we've agreed on the API.

Reply:
To use a stuct vs. a class must be decided now, because altering that is a breaking change, which is not allowed later. It is not the time of destroying that matters here, rather it is specifically the Garbage Collector that causes problems. A struct is actually not affected by the GC, because that is a value type, in contrast with a class, which is a reference type. A struct is a good approach here, because that is automatically destroyed when it goes out of scope, but a class is just registered for later destoying by the GC. If it is decided to use a class instead, your approach of reusing the same class across all calls must be used.

There were extended efforts in OpenRails to hunt down class instance creations in the game loop in the past, to minimize the triggering of the GC. The problem with the GC is that when it runs, it hangs the game loop for some time, thus causes micro-stuttering. Ideally there should be no new class instance creations in the game loop at all, all needed classes must be created at the game initialization time, and never destroyed/disposed. I learned this from James @twpol, he might jump in here either to deny or confirm this.

Reply:
Technically some scripts could continue to work if the names and syntax remained identical - as in `NextGenericSignalInfo(...).Aspect` - but yes, I realize now that if you do anything involving the type declaration, that would introduce the possibility of breakage.

Okay, so stick with structs, but I suggest adding the `readonly` keyword to keep their fields immutable.

Reply:
OK

Reply:
Setting the Struct as readonly generates a compile error in the TrainControlSystem.cs statements where I set their values.

Reply:
It's possible if you give the struct a constructor. See my [attached patch](https://github.com/openrails/openrails/files/5502929/use-immutable-structs.patch.txt).

Reply:
OK, I'll apply it.

Reply:
Ryan, your patch does work for direct calls to NextGenericSignalFeatures(), but doesn't work for calls to NextGenericSignalItem().
Apparently from NextGenericSignalItem() it enters NextGenericSignalFeatures(), returns correctly to NextGenericSignalItem(), but instead of returning from it it enters again in the Return statement of NextGenericSignalFeatures(), and exiting to it into NextGenericSignalItem the parameter values are lost.

Reply:
I didn't expect you to store the result of a function call in an instance variable. Try this [revised patch](https://github.com/openrails/openrails/files/5503635/immutable-structs-2.patch.txt), please.

Reply:
Patch OK, thanks. I've just pushed the new PR release including it.